db/inst/import-try/03.create.try.sites.R

```diff
@@ -0,0 +1,105 @@
+hc.latlon <- hclust(dist(latlon.unique))
+clusters.latlon <- cutree(hc.latlon, h = radius)
+latlon.unique[, try.site.id := paste0("TRY_SITE_", clusters.latlon)]
+
+# This computes the site centroid, which is just the mean latitude and longitude for all points in the site
+latlon.unique[, c("site.Latitude", "site.Longitude") := list(mean(Latitude, na.rm=TRUE),
+                                                             mean(Longitude, na.rm=TRUE)), by=try.site.id]
+latlon.unique <- latlon.unique[!(is.na(Latitude) | is.na(Longitude))]
+
+# Merge back into try.latlon.cast
+setkey(latlon.unique, latlon.id)
+setkey(try.latlon.cast, latlon.id)
+try.latlon.merge <- try.latlon.cast[latlon.unique[,list(latlon.id, try.site.id, 
+                                                        site.Latitude, site.Longitude)]]
+
+# d. Append location.id and site.id to full data set by merging on ObservationID
+setkey(try.latlon.merge, ObservationID)
+setkey(try.dat, ObservationID)
+try.dat <- try.dat[try.latlon.merge[,list(ObservationID, latlon.id, try.site.id,
+                                          site.Latitude, site.Longitude)]]
+
+# e. Create data.table for site creation:
+#   sitename = "TRY_SITE_<site.id>"
+#   notes = "TRY_DATASETS = <dataset IDs>"
+#   geometry = ST_GeomFromText('POINT(lat, lon)', 4263)
+try.sites <- try.dat[, paste("TRY_DATASETS =", paste(unique(DatasetID), collapse=" ")),
+                     by=list(try.site.id, site.Latitude, site.Longitude)]
+setnames(try.sites, "V1", "notes")
+try.sites[, bety.site.id := as.character(NA)]
+bety.site.index <- which(names(try.sites) == "bety.site.id")
+
+# f. Loop over rows... 
+radius.query.string <- 'SELECT id, sitename, ST_Y(ST_Centroid(geometry)) AS lat, ST_X(ST_Centroid(geometry)) AS lon, ST_Distance(ST_Centroid(geometry), ST_SetSRID(ST_MakePoint(%2$f, %1$f), 4326)) as distance FROM sites WHERE ST_Distance(ST_Centroid(geometry), ST_SetSRID(ST_MakePoint(%2$f, %1$f), 4326)) <= %3$f'
+insert.query.string <- "INSERT INTO sites(sitename,notes,geometry,user_id,created_at,updated_at) VALUES('%s','%s',ST_Force3D(ST_SetSRID(ST_MakePoint(%f, %f), 4326)),'%s', NOW(), NOW() ) RETURNING id;"
+
+message("Looping over sites and adding to BETY")
+pb <- txtProgressBar(0, nrow(try.sites), style=3)
+for(r in 1:nrow(try.sites)){
+  # Check site centroid against BETY.
+  site.lat <- try.sites[r, site.Latitude]
+  site.lon <- try.sites[r, site.Longitude]
+  search.df <- try(db.query(sprintf(radius.query.string, site.lat, site.lon, radius), con))
+  if(class(search.df) == "try-error"){
+    warning("Error querying database.")
+    next
+  }
+  if(nrow(search.df) > 0){
+    search.df <- search.df[order(search.df$distance),]
+    rownames(search.df) <- 1:nrow(search.df)
+    newsite <- FALSE
+    ## Select closest existing site automatically, if it's within the radius.
+    bety.site.id <- as.character(search.df[1,"id"])
+    ## TODO: Alternative is to select sites manually at each step, as implemented below.
```

Comment:
@dlebauer Here is the block for manually adding sites. For each site with matches within the given radius, it prompts you with a menu of BETY sites sorted by distance and lets you choose one, or to create a new site anyway.